x/stakeibc/keeper/lsm_test.go

```diff
@@ -202,9 +206,86 @@ func (s *KeeperTestSuite) TestGetValidatorFromLSMTokenDenom() {
 	s.Require().ErrorContains(err, "validator (cosmosvaloperXXX) is not registered in the Stride validator set")
 
 	// Pass in a validator that has a slash query in flight - it should fail
-	validatorsWithPendingQuery := []*types.Validator{{Address: valAddress, SlashQueryInProgress: true}}
-	_, err = s.App.StakeibcKeeper.GetValidatorFromLSMTokenDenom(denom, validatorsWithPendingQuery)
+	validatorWithSlashQuery := []*types.Validator{{
+		Address:                    valAddress,
+		SlashQueryInProgress:       true,
+		InternalSharesToTokensRate: sdk.OneDec(),
+	}}
+	_, err = s.App.StakeibcKeeper.GetValidatorFromLSMTokenDenom(denom, validatorWithSlashQuery)
 	s.Require().ErrorContains(err, "validator cosmosvaloperXXX was slashed")
+
+	// Pass in a validator with an uninitialized exchange rate - it should fail
+	validatorWithoutExchangeRate := []*types.Validator{{
+		Address:              valAddress,
+		SlashQueryInProgress: false,
+	}}
+	_, err = s.App.StakeibcKeeper.GetValidatorFromLSMTokenDenom(denom, validatorWithoutExchangeRate)
+	s.Require().ErrorContains(err, "validator cosmosvaloperXXX exchange rate is not known")
+}
+
+func (s *KeeperTestSuite) TestCalculateLSMStToken() {
+	testCases := []struct {
+		name                  string
+		liquidStakedShares    sdkmath.Int
+		validatorExchangeRate sdk.Dec
+		redemptionRate        sdk.Dec
+		expectedStAmount      sdkmath.Int
+	}{
+		// stTokenAmount = liquidStakedShares * validatorExchangeRate / redemptionRate
+		{
+			name:                  "one exchange rate and redemption rate",
+			liquidStakedShares:    sdkmath.NewInt(1000),
+			validatorExchangeRate: sdk.OneDec(),
+			redemptionRate:        sdk.OneDec(),
+			expectedStAmount:      sdkmath.NewInt(1000),
+		},
+		{
+			name:                  "one exchange rate, non-one redemption rate",
+			liquidStakedShares:    sdkmath.NewInt(1000),
+			validatorExchangeRate: sdk.OneDec(),
+			redemptionRate:        sdk.MustNewDecFromStr("1.25"),
+			expectedStAmount:      sdkmath.NewInt(800), // 1000 * 1 / 1.25 = 800
+		},
+		{
+			name:                  "non-one exchange rate, one redemption rate",
+			liquidStakedShares:    sdkmath.NewInt(1000),
+			validatorExchangeRate: sdk.MustNewDecFromStr("0.75"),
+			redemptionRate:        sdk.OneDec(),
+			expectedStAmount:      sdkmath.NewInt(750), // 1000 * 0.75 / 1
+		},
+		{
+			name:                  "non-one exchange rate, non-one redemption rate",
+			liquidStakedShares:    sdkmath.NewInt(1000),
+			validatorExchangeRate: sdk.MustNewDecFromStr("0.75"),
+			redemptionRate:        sdk.MustNewDecFromStr("1.25"),
+			expectedStAmount:      sdkmath.NewInt(600), // 1000 * 0.75 / 1.25 = 600
```

Comment:
If the user stakes 1 share, the val exchrate is < 1 and the RR is >= 1, do they get 0 stTokens? Ik it's unlikely anyone would input 1e-6 on the frontend, but to sanity check, would that case throw an error?

Reply:
good thought. It would convert to 0 stTokens but then the liquid stake would fail [here](https://github.com/Stride-Labs/stride/blob/lsm/x/stakeibc/keeper/msg_server_lsm_liquid_stake.go#L78)